core/src/main/java/org/springframework/security/oauth2/server/authorization/OAuth2Authorization.java

```diff
@@ -15,17 +15,79 @@
  */
 package org.springframework.security.oauth2.server.authorization;
 
-import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
 
+import java.util.Collections;
 import java.util.Map;
+import java.util.Objects;
 
 /**
+ * Represents a collection of attributes which describe an OAuth 2.0 authorization context.
+ *
  * @author Joe Grandja
+ * @author Krisztian Toth
  */
-public class OAuth2Authorization {
-	private String registeredClientId;
-	private String principalName;
-	private OAuth2AccessToken accessToken;
-	private Map<String, Object> attributes;
+public abstract class OAuth2Authorization implements TokenContainer {
```

Comment:
Please remove `abstract`. This domain object does not need to be extended at the moment.

Reply:
Thanks for the review @jgrandja. I wasn't exactly sure what would be the best way to derive the `TokenType` from an `OAuth2Authorization` object. If i am correct there is either a `code` entry in the `attributes` map or an `OAuthAccessToken` field in this class - what do you think would be the best way to get the token and the TokenType attributes in the `InMemoryOAuth2AuthorizationService`? As I can see, we either 
* do that in this class with a `TokenType` field that is assigned to when the object is created:
```java
	public OAuth2Authorization(String registeredClientId, String principalName,
			@Nullable OAuth2AccessToken accessToken, @Nullable Map<String, Object> attributes) {
		// ...
                this.tokenType = // AUTHORIZATION_CODE if accessToken is null and code is set in attributes, 
                                 // ACCESS_TOKEN otherwise, throw exception if none
	}
```
This feels really clunky as this class would have to be modified every time there is a new TokenType (and TokenType also has a public constructor). We would also have to fetch the token from different sources - if I am not mistaken - depending on the token type, so there should also be a field for the token itself.
* or we create an external class with a method signature something along the lines of
```java
public TokenAndType getTokenAndTypeFrom(OAuth2AccessToken accessToken) { ... }
```
But perhaps I am way off, that could also very well be. Could you please advise? Thank you!

Reply:
@watsta This is what I envision.

Given the current design:

```java
public class OAuth2Authorization {
	private String registeredClientId;
	private String principalName;
	private OAuth2AccessToken accessToken;
	private Map<String, Object> attributes;

	...

}

public interface OAuth2AuthorizationService {

	void save(OAuth2Authorization authorization);

	OAuth2Authorization findByTokenAndTokenType(String token, TokenType tokenType);

}

```

When the Authorization Endpoint #66 responds with the [Authorization Response](https://tools.ietf.org/html/rfc6749#section-4.1.2), the _in-flight_ `OAuth2Authorization` should be `OAuth2AuthorizationService.save()` as such:

```java
public class OAuth2Authorization {
	private String registeredClientId;     // auto-assigned identifier
	private String principalName;     // current `Authentication.getName()`
	private OAuth2AccessToken accessToken;     // null since access token is not available yet
	private Map<String, Object> attributes;     // `code` response parameter assigned to well-known key

	...

}

```

Now when the client calls the Token Endpoint #67,  the [Access Token Request ](https://tools.ietf.org/html/rfc6749#section-4.1.3) will include the `code` parameter issued by #66, which can be looked up using:

```java
OAuth2AuthorizationService.findByTokenAndTokenType(code, TokenType.AUTHORIZATION_CODE)
```

The implementation would use the well-known key (eg. `OAuth2AuthorizationAttributeNames.AUTHORIZATION_CODE`) to lookup in `OAuth2Authorization.attributes` for each `OAuth2Authorization`, and if found matching `code`, return it for further validation by the caller.

Assuming the caller #67 validates successfully, then the `OAuth2Authorization` is updated with the `OAuth2AccessToken` resulting in this state at the end of the grant flow.

```java
public class OAuth2Authorization {
	private String registeredClientId;     // previously assigned
	private String principalName;     // previously assigned
	private OAuth2AccessToken accessToken;     // set the granted access token
	private Map<String, Object> attributes;    // previously assigned

	...

}
```

Does this clarify things?

By the way, have you had a chance to read the [Getting Started](https://github.com/spring-projects-experimental/spring-authorization-server#getting-started)? If not, then I strongly encourage you to read the [OAuth 2.0 Authorization Framework](https://tools.ietf.org/html/rfc6749) to help clarify things further.

Reply:
Thank you for taking your time to clarify @jgrandja , it certainly does help. My confusion comes from the fact that for different `TokenType`s we will have to use different logic to save the authorization object. 

In your example, when #66 calls `#save(OAuth2Authorization)` we want to know what the derived `TokenType` of the object is so we can use the lookup method `#findByTokenAndTokenType(token, TokenType)`. This is obviously not a problem, when we want to save the authorization we can just grab the `code` entry from the `attributes` map. When it becomes a problem is when you want to save authorizations with different `TokenType`s than `AUTHORIZATION_CODE`. My initial question was, where should this derivation logic exist? This is what I wanted to solve with the current implementation. 
Should I implement the in-memory service which always saves the authorization object with `AUTHORIZATION_CODE` token type for now?

I think I understand the role of the `OAuth2AuthorizationService` in the application and I am familiar with the OAuth framework to an extent, but I will make sure to read through the specifications and the resources in the Getting Started page, thanks!.

Reply:
> Should I implement the in-memory service which always saves the authorization object with AUTHORIZATION_CODE token type for now

The `OAuth2AccessToken` (only available in 2nd phase of flow) will also be `save(OAuth2Authorization)` along with the `AUTHORIZATION_CODE` in the 1st phase of the flow. So as far as looking up the `OAuth2AccessToken` using `OAuth2AuthorizationService.findByTokenAndTokenType(code, TokenType.ACCESS_TOKEN)`, we simply search for it via the `OAuth2AccessToken` member.